### ISSUE TITLE
Cursor NULL / Collect all data

### ISSUE THREAD

**Issue by haasaah** [hmbouk/Monday-paginated-board-PBI-query#2]
Hi **Howard**,

First, I would like to thank you, your code / solution got me out of limbo with this new Monday API.

I managed to operationalize all the logic, using "CURSOR", exactly with your code (GetBoard_1stPage / GetBoard_RemainingPages)

Again, thank you very much.

Now, I need help (if I can) to join the tables where CURSOR = NULL. I created the code below, considering scanning all queries in a loop, as I have several BOARDS on Monday with more than 5000 rows, , and which grow dynamically, very quickly.

```
let
  BoardID = XXXXXX,
  Combined_Table = List.Generate(
        () => [i = 0, cursor = "cursor", result = #table({"cursor"}, {{"NoMorePages"}})],
        each [i] <= 100 and [cursor] <> "NoMorePages",
        each [
            i = [i] + 1,
            page_table = if [i] = 0 then GetBoard_1stPage_teste(BoardID) else GetBoard_RemainingPages_teste([cursor]),
            page_table_updated = Table.ReplaceValue(page_table, null, "NoMorePages", Replacer.ReplaceValue, {"cursor"}),
            cursor = page_table_updated{0}[cursor],
            result = page_table_updated
        ],
        each [result]
    ),
  Combined_Table_COMPLETO = Table.Combine(Combined_Table)
 
in
  #"Combined_Table_COMPLETO"
```

In the example below (considering the query limit is 4 (for example)):

![image](https://github.com/hmbouk/Monday-paginated-board-PBI-query/assets/12752872/04091674-7878-4977-97cd-3852477b894f)

   
1. search is performed using the "GetBoard_1stPage_teste" function, to collect the first CURSOR identifier (HASH 1);
2. the second query is made, with the function "GetBoard_RemainingPages_teste", using CURSOR = HASH 1, and returning CURSOR = HASH 11 for the next query;
3. the third query is made, with the function "GetBoard_RemainingPages_teste", using CURSOR = HASH 11, and returning CURSOR = HASH 111 for the next query;
4. the fourth query is made, with the "GetBoard_RemainingPages_teste" function, using CURSOR = HASH 111, and returning CURSOR = HASH 1111 for the next query;
5. the fifth query is made, with the "GetBoard_RemainingPages_teste" function, using CURSOR = HASH 1111, and returning CURSOR = HASH 11111 for the next query;
6. the sixth query is made, with the function "GetBoard_RemainingPages_teste", using CURSOR = HASH 11111, and returning CURSOR = null, when there is no next cursor for query.

After the end of the script, the data from the last query of CURSOR = HASH 11111, which returns CURSOR = null, is not combined in the query.

If I force it, and apply the HASH 11111 directly in the "RemainingPages_teste" function, the data is shown.

In a real example: in a query to a BOARD of 2172 lines, with a limit of 100, the code runs perfectly, however, when analyzing the table, only 2101 are combined (+ "NoMorePages" line), disregarding the last 72 lines of information.

I've already tried changing conditions, doing several validation tests, but no option, it made me append the data with CURSOR = NULL.

Please, could you give me some help? anything I missed?

Thank you very much, and much success!

**Comment by haasaah** [comment]
All,

I adjusted a temporary code (I understand that it is not the best solution), but the idea is to register the last searched cursor, and perform the search with cursor=null and match the existing table.

```
let
    BoardID = 4139784827,
    
    Combined_Table = List.Generate(
        () => [i = 0, cursor = "cursor", result = #table({"cursor"}, {{"NoMorePages"}})],
        each [i] <= 100 and [cursor] <> "NoMorePages",
        each [
            i = [i] + 1,
            page_table = if [i] = 0 then GetBoard_1stPage_teste(BoardID) else GetBoard_RemainingPages_teste([cursor]),
            page_table_updated = Table.ReplaceValue(page_table, null, "NoMorePages", Replacer.ReplaceValue, {"cursor"}),
            cursor = page_table_updated{0}[cursor],
            result = page_table_updated
        ],
        each [result]
    ),
    Combined_Table1 = Table.Combine(Combined_Table),

    last_cursor = List.Last(Table.Column(Combined_Table1, "cursor")),
    last_page = GetBoard_RemainingPages_teste(last_cursor),
    
    #"Linhas filtradas" = Table.SelectRows(Table.Combine({Combined_Table1, last_page}), each [ID] <> null and [ID] <> "")
in
    #"Linhas filtradas"
```

Please, if you have an idea to improve/adjust the code, I am at your disposal.

Tks !

**Comment by hmbouk** [comment]
> Hi **Howard**,
> 
> First, I would like to thank you, your code / solution got me out of limbo with this new Monday API.
> 
> I managed to operationalize all the logic, using "CURSOR", exactly with your code (GetBoard_1stPage / GetBoard_RemainingPages)
> 
> Again, thank you very much.
> 
> Now, I need help (if I can) to join the tables where CURSOR = NULL. I created the code below, considering scanning all queries in a loop, as I have several BOARDS on Monday with more than 5000 rows, , and which grow dynamically, very quickly.
> 
> ```
> let
>   BoardID = XXXXXX,
>   Combined_Table = List.Generate(
>         () => [i = 0, cursor = "cursor", result = #table({"cursor"}, {{"NoMorePages"}})],
>         each [i] <= 100 and [cursor] <> "NoMorePages",
>         each [
>             i = [i] + 1,
>             page_table = if [i] = 0 then GetBoard_1stPage_teste(BoardID) else GetBoard_RemainingPages_teste([cursor]),
>             page_table_updated = Table.ReplaceValue(page_table, null, "NoMorePages", Replacer.ReplaceValue, {"cursor"}),
>             cursor = page_table_updated{0}[cursor],
>             result = page_table_updated
>         ],
>         each [result]
>     ),
>   Combined_Table_COMPLETO = Table.Combine(Combined_Table)
>  
> in
>   #"Combined_Table_COMPLETO"
> ```
> 
> In the example below (considering the query limit is 4 (for example)):
> 
> ![image](https://private-user-images.githubusercontent.com/12752872/295958354-04091674-7878-4977-97cd-3852477b894f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUwOTU4OTAsIm5iZiI6MTcwNTA5NTU5MCwicGF0aCI6Ii8xMjc1Mjg3Mi8yOTU5NTgzNTQtMDQwOTE2NzQtNzg3OC00OTc3LTk3Y2QtMzg1MjQ3N2I4OTRmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEyVDIxMzk1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM0NGEyZDNlMmIzMWU4YjA3YjNjNzBhODQ2MjlhODUzMGExZjc2NjM5NzgzNGIyYzc3NjJlMWU5MGJlYjA2ZWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.fMf6BpbyDhflfUGXqJq7suIylySiZfQqBqsZ1j3WCJQ)
> 
> 1. search is performed using the "GetBoard_1stPage_teste" function, to collect the first CURSOR identifier (HASH 1);
> 2. the second query is made, with the function "GetBoard_RemainingPages_teste", using CURSOR = HASH 1, and returning CURSOR = HASH 11 for the next query;
> 3. the third query is made, with the function "GetBoard_RemainingPages_teste", using CURSOR = HASH 11, and returning CURSOR = HASH 111 for the next query;
> 4. the fourth query is made, with the "GetBoard_RemainingPages_teste" function, using CURSOR = HASH 111, and returning CURSOR = HASH 1111 for the next query;
> 5. the fifth query is made, with the "GetBoard_RemainingPages_teste" function, using CURSOR = HASH 1111, and returning CURSOR = HASH 11111 for the next query;
> 6. the sixth query is made, with the function "GetBoard_RemainingPages_teste", using CURSOR = HASH 11111, and returning CURSOR = null, when there is no next cursor for query.
> 
> After the end of the script, the data from the last query of CURSOR = HASH 11111, which returns CURSOR = null, is not combined in the query.
> 
> If I force it, and apply the HASH 11111 directly in the "RemainingPages_teste" function, the data is shown.
> 
> In a real example: in a query to a BOARD of 2172 lines, with a limit of 100, the code runs perfectly, however, when analyzing the table, only 2101 are combined (+ "NoMorePages" line), disregarding the last 72 lines of information.
> 
> I've already tried changing conditions, doing several validation tests, but no option, it made me append the data with CURSOR = NULL.
> 
> Please, could you give me some help? anything I missed?
> 
> Thank you very much, and much success!

Good catch @haasaah.  I'll have a look and fix this in my query and then upload the changes.  Sorry for the delay, I have had the flu for the last few days.

**Comment by haasaah** [comment]
> Good catch @haasaah. I'll have a look and fix this in my query and then upload the changes. Sorry for the delay, I have had the flu for the last few days.

@hmbouk , Thank you, I'm at your disposal for any ideas.

Good recovery, get better.
